transaction/src/builder/manifest_namer.rs

```diff
@@ -0,0 +1,733 @@
+        Self(GLOBAL_INCREMENTER.fetch_add(1, core::sync::atomic::Ordering::Acquire))
+    }
+}
+
+/// This exposes a shared stateful core between a ManifestNamer and its corresponding Registrar.
+/// This core is wrapped in a Rc<RefCell> for sharing between the two.
+#[derive(Default)]
+struct ManifestNamerCore {
+    namer_id: ManifestNamerId,
+    id_allocator: ManifestIdAllocator,
+    named_buckets: IndexMap<String, ManifestObjectState<ManifestBucket>>,
+    named_proofs: IndexMap<String, ManifestObjectState<ManifestProof>>,
+    named_addresses: NonIterMap<String, ManifestObjectState<ManifestAddress>>,
+    named_address_reservations: NonIterMap<String, ManifestObjectState<ManifestAddressReservation>>,
+    object_names: ManifestObjectNames,
+}
+
+impl ManifestNamerCore {
+    pub fn new_named_bucket(&mut self, name: impl Into<String>) -> NamedManifestBucket {
+        let name = name.into();
+        let old_entry = self
+            .named_buckets
+            .insert(name.clone(), ManifestObjectState::Unregistered);
+        if old_entry.is_some() {
+            panic!("You cannot create a new bucket with the same name \"{name}\" multiple times");
+        }
+        NamedManifestBucket {
+            namer_id: self.namer_id,
+            name,
+        }
+    }
+
+    pub fn new_named_bucket_collision_free(
+        &mut self,
+        prefix: &str,
+    ) -> (String, NamedManifestBucket) {
+        for name_counter in 1..u32::MAX {
+            let name = if name_counter == 1 {
+                prefix.to_string()
+            } else {
+                format!("{prefix}_{name_counter}")
+            };
+            if !self.named_buckets.contains_key(&name) {
+                let new = self.new_named_bucket(&name);
+                return (name, new);
+            }
+        }
+        panic!("Did not resolve a name");
+    }
+
+    pub fn resolve_named_bucket(&self, name: impl AsRef<str>) -> ManifestBucket {
+        match self.named_buckets.get(name.as_ref()) {
+            Some(ManifestObjectState::Present(bucket)) => bucket.clone(),
+            Some(ManifestObjectState::Consumed) => panic!("Bucket with name \"{}\" has already been consumed", name.as_ref()),
+            _ => panic!("You cannot use a bucket with name \"{}\" before creating it with `new_bucket` and passing that into the relevant instruction in the manifest builder", name.as_ref()),
+        }
+    }
+
+    /// This is intended for registering a bucket name to an allocated identifier, as part of processing a manifest
+    /// instruction which creates a bucket.
+    pub fn register_bucket(&mut self, new: NamedManifestBucket) {
+        if self.namer_id != new.namer_id {
+            panic!("NewManifestBucket cannot be registered against a different ManifestNamer")
+        }
+        let new_bucket = self.id_allocator.new_bucket_id();
+        match self.named_buckets.get_mut(&new.name) {
+            Some(allocated @ ManifestObjectState::Unregistered) => {
```

Comment:
That triggers the code path `Some(_) => unreachable!("NewManifestBucket was somehow registered twice")`

But indeed - it's unreachable, because to get here, you need to have created two `NamedManifestBucket`s with `new_bucket` (because they're not clone) - and we get the user-facing error message there about a duplicate bucket name.